javalib/src/main/scala/java/util/AbstractMap.scala

```diff
@@ -0,0 +1,199 @@
+package java.util
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+
+object AbstractMap {
+
+  private def entryEquals[K, V](entry: Map.Entry[K, V], other: Any): Boolean = {
+    other match {
+      case other: Map.Entry[_, _] =>
+        entry.getKey == other.getKey && entry.getValue == other.getValue
+      case _ => false
+    }
+  }
+
+  private def entryHashCode[K, V](entry: Map.Entry[K, V]): Int = {
+    val keyHash =
+      if (entry.getKey == null) 0
+      else entry.getKey.hashCode
+    val valueHash =
+      if (entry.getValue == null) 0
+      else entry.getValue.hashCode
+
+    keyHash ^ valueHash
+  }
+
+  class SimpleEntry[K, V](private var _key: K, private var _value: V) extends Map.Entry[K, V] with Serializable {
+
+    def this(entry: Map.Entry[K, V]) =
+      this(entry.getKey, entry.getValue)
+
+    override def equals(o: Any): Boolean =
+      entryEquals(this, o)
+
+    def getKey(): K = _key
+
+    def getValue(): V = _value
+
+    override def hashCode(): Int = entryHashCode(this)
+
+    def setValue(value: V): V = {
+      val oldValue = _value
+      _value = value
+      oldValue
+    }
+  }
+
+  class SimpleImmutableEntry[K, V](key: K, value: V) extends Map.Entry[K, V] with Serializable {
+    def this(entry: Map.Entry[K, V]) =
+      this(entry.getKey, entry.getValue)
+
+    override def equals(o: Any): Boolean =
+      entryEquals(this, o)
+
+    def getKey(): K = key
+
+    def getValue(): V = value
+
+    override def hashCode(): Int = entryHashCode(this)
+
+    def setValue(value: V): V = throw new UnsupportedOperationException()
+  }
+}
+
+abstract class AbstractMap[K, V]() extends java.util.Map[K, V] { self =>
+
+  @tailrec
+  private def findFirst(i: Iterator[Map.Entry[K, V]])(f: (Map.Entry[K, V]) => Boolean,
+      fi: (Iterator[Map.Entry[K, V]]) => Unit = { _ => () }): Map.Entry[K, V] = {
+    if (!i.hasNext) null
+    else {
+      val o = i.next
+      if (f(o)) {
+        fi(i)
+        o
+      } else {
+        findFirst(i)(f)
+      }
+    }
+  }
+
+  @tailrec
+  private def onAll[IK, IV](i: Iterator[Map.Entry[IK, IV]])(f: (Map.Entry[IK, IV]) => Unit): Unit = {
+    if (i.hasNext) {
+      f(i.next)
+      onAll(i)(f)
+    }
+  }
+
+  def size(): Int = entrySet.size
+
+  def isEmpty(): Boolean = size == 0
+
+  def containsValue(value: Any): Boolean =
+    entrySet.iterator.exists(value == _.getValue)
+
+  def containsKey(key: Any): Boolean =
+    get(key) != null
```

Comment:
This is not correct, as key would return `true` if the map contains an entry whose `value eq null`, regardless of its key.

It should be `iterator.exists(entry => sameValue(entry, key))`